prelude.dx

```diff
@@ -405,3 +405,23 @@ def loadTile (t : Tile n (Fin VectorWidth)) (arr : n=>Real) : VectorReal =
               arr.(t +> UNSAFEFromOrdinal idx 1)
               arr.(t +> UNSAFEFromOrdinal idx 2)
               arr.(t +> UNSAFEFromOrdinal idx 3))
+
+'Numerical utilities
+
+def logsumexp (x: n=>Real) : Real =
+  m = maximum x
+  m + (log $ sum for i. exp (x.i - m))
+
+def logaddexp (x:Real) (y:Real) : Real =
```

Comment:
I wonder if for this one it wouldn't be better to define a newtype and define this as its `(+)` operation? WDYT?

Reply:
I have been wanting such a number type for quite a while, but never enough to actually implement it.
Also would get to define `*`, and `pow` very nicely.

Reply:
I didn't realize until you suggested it that this might be possible in Dex as it stands right now, but now I want to try it out!

It would certainly make the CTC example much clearer, and would have made it easier to write.  I first wrote the whole thing in terms of (+, *), then converted it all to (logaddexp, +) by hand.  I'll try writing this up in the next day or two.

Even if it has unforeseen problems, I think this is exactly the sort of thing that we should be using Dex to explore.  I really like it when the main loop of the real code can be made to look like the pseudocode.